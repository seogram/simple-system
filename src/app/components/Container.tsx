import React from "react";
import { useState, useEffect } from "react";
import { useSearchParams } from "next/navigation";
import SearchBox from "./SearchBox";
import ResultCard from "./ResultCard";
import Wrapper from "./Wrapper";

const Container = () => {
  const searchParams = useSearchParams();
  const [searchTerm, setSearchTerm] = useState<string | undefined>();

  const query = searchParams.get("user");
  useEffect(() => {
    if (query) {
      setSearchTerm(query);
    }
  }, [query]);

return (
    <Wrapper>
    <SearchBox />
    <ResultCard searchTerm={searchTerm} />
  </Wrapper>
)
};

export default Container;
