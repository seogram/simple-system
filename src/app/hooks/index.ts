export {  useUserData , useUserRepo } from "./useUserData";
