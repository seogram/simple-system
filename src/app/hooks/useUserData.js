import { useQuery } from "react-query";
// ----------------------------------------------------------------------
const baseUrl = "https://api.github.com";

const request = async (url) => {
  try {
    const response = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    const res = await response.json();
    return res;
  } catch (error) {
    console.log("error for", error);
    return {
      res: [],
    };
  }
};
const getGithubUserData = async (query) => {
  if (!query) return;

  const params = {
    q: query,
    per_page: 5,
  };
  const qs = "?" + new URLSearchParams(params).toString();
  const url = `${baseUrl}/search/users${qs}`;
  const response = await request(url);
  return response?.items;
};

const getGithubUserRepos = async (username) => {
  if (!username) return;
  const url = `${baseUrl}/users/${username}/repos`;

  const response = await request(url);
  return response;
};

export function useUserData(username) {
  const { isFetching, isError, data, error } = useQuery(
    `githubuser-${username}`,
    () => getGithubUserData(username),
    {
      enabled: Boolean(username),
      refetchOnWindowFocus: false,
    }
  );
  return {
    isFetching,
    isError,
    data,
    error,
  };
}

export function useUserRepo({username,enabled}) {
  const { isFetching, isError, data, error } = useQuery(
    `githubRepo-${username}`,
    () => getGithubUserRepos(username),
    {
      enabled: Boolean(username && enabled),
      refetchOnWindowFocus: false,
    }
  );
  return {
    isFetching,
    isError,
    data,
    error,
  };
}
