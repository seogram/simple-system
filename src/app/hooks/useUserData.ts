import { useQuery } from "react-query";
// ----------------------------------------------------------------------
const BASE_URL = "https://api.github.com";

const logSentry = () => {
  //Log in Sentry or other log systems
};
const request = async (url: string) => {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      const errorCode = response.status;
      throw new Error(`An error occurred: ${errorCode}`);
    }
    return response.json();
  } catch (error) {
    logSentry();
    throw error;
  }
};
const getGithubUserData = async (query?: string) => {
  if (!query) return;

  const params = {
    q: query,
    per_page: "5",
  };
  const qs = "?" + new URLSearchParams(params).toString();
  const url = `${BASE_URL}/search/users${qs}`;
  const response = await request(url);
  return response?.items;
};

const getGithubUserRepos = async (username?: string) => {
  if (!username) return;
  const url = `${BASE_URL}/users/${username}/repos`;

  const response = await request(url);
  return response;
};

export function useUserData(username?: string) {
  const { isFetching, isError, data, error } = useQuery(
    `githubuser-${username}`,
    () => getGithubUserData(username),
    {
      enabled: Boolean(username),
      refetchOnWindowFocus: false,
    }
  );
  return {
    isFetching,
    isError,
    data,
    error,
  };
}

export function useUserRepo({
  username,
  enabled,
}: {
  username?: string;
  enabled: boolean;
}) {
  const { isFetching, isError, data, error } = useQuery(
    `githubRepo-${username}`,
    () => getGithubUserRepos(username),
    {
      enabled: Boolean(username && enabled),
      refetchOnWindowFocus: false,
    }
  );
  return {
    isFetching,
    isError,
    data,
    error,
  };
}
