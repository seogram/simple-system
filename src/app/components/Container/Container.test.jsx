import React from "react";
import { useSearchParams } from "next/navigation";
import { render } from "@testing-library/react";
import Container from "./Container";
import { useUserData, useUserRepo } from "../../hooks/useUserData";
import { QueryClient, QueryClientProvider } from "react-query";

jest.mock("../../hooks/useUserData");

const mockUserRepoData = useUserRepo;

jest.mock("../../hooks/useUserData");

const mockUserData = useUserData;

jest.mock("next/navigation", () => {
  return {
    useRouter: jest.fn(),
    useSearchParams: jest.fn(),
  };
});

describe("Container Component ", () => {
  let queryClient;
  beforeEach(() => {
    queryClient = new QueryClient();
  });

  test("should render without crashing when there is no searchTerm", async () => {
    useSearchParams.mockReturnValue({ get: jest.fn() });

    mockUserRepoData.mockReturnValue({
      data: undefined,
      isFetching: false,
      isError: false,
    });

    mockUserData.mockReturnValue({
      githubUserRepo: undefined,
      isFetching: false,
      isError: false,
    });

    const container = render(
      <QueryClientProvider client={queryClient}>
        <Container />
      </QueryClientProvider>
    );

    expect(container).toMatchSnapshot();
  });

  test("should render without crashing when there is a query param : user=", async () => {
    useSearchParams.mockReturnValue({ get: () => "testUser" });

    mockUserRepoData.mockReturnValue({
      data: undefined,
      isFetching: false,
      isError: false,
    });

    mockUserData.mockReturnValue({
      githubUserRepo: undefined,
      isFetching: true,
      isError: false,
    });

    const { getByText } = render(
      <QueryClientProvider client={queryClient}>
        <Container />
      </QueryClientProvider>
    );

    const loadingText = getByText("Loading...");
    expect(loadingText).toBeInTheDocument();
  });

  test("should render correctly when data is fetched", async () => {
    useSearchParams.mockReturnValue({ get: () => "testUser" });

    mockUserRepoData.mockReturnValue({
      data: undefined,
      isFetching: false,
      isError: false,
    });

    mockUserData.mockReturnValue({
      githubUserRepo: [{ id: 1, login: "test", repos_url: "/test" }],
      isFetching: false,
      isError: false,
    });

    const { getByText } = render(
      <QueryClientProvider client={queryClient}>
        <Container />
      </QueryClientProvider>
    );

    expect(getByText("Showing users for testUser :")).toBeInTheDocument();
  });
});
