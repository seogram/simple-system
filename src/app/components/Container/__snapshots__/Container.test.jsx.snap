// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Component  should render without crashing when there is no searchTerm 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-kms3ws-MuiPaper-root"
      >
        <div
          class="MuiBox-root css-1hc4zkj"
        >
          <div
            class="MuiStack-root css-163ljxj-MuiStack-root"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-1dfpgwu-MuiInputBase-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                name="username"
                placeholder="Username..."
                type="text"
                value=""
              />
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-s064ox-MuiButtonBase-root-MuiButton-root"
              data-testid="searchButton"
              disabled=""
              name="search"
              tabindex="-1"
              type="button"
            >
              Search
            </button>
          </div>
        </div>
        <div
          class="MuiStack-root css-wy8e47-MuiStack-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-0"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-kms3ws-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-1hc4zkj"
      >
        <div
          class="MuiStack-root css-163ljxj-MuiStack-root"
        >
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth css-1dfpgwu-MuiInputBase-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
              name="username"
              placeholder="Username..."
              type="text"
              value=""
            />
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-s064ox-MuiButtonBase-root-MuiButton-root"
            data-testid="searchButton"
            disabled=""
            name="search"
            tabindex="-1"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
      <div
        class="MuiStack-root css-wy8e47-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-0"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
